File: packages/b2c-tooling-sdk/src/auth/oauth-jwt.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
/**
 * JWT Bearer OAuth authentication strategy (RFC 7523).
 *
 * Implements OAuth 2.0 JWT Bearer Token flow for Account Manager authentication.
 * Uses client certificate/key pair instead of client secret for enhanced security.
 *
 * @module auth/oauth-jwt
 */
import * as crypto from 'node:crypto';
import * as fs from 'node:fs';
import type {AuthStrategy, FetchInit, AccessTokenResponse} from './types.js';
import {getLogger} from '../logging/logger.js';
import {
  getOAuthCacheKey,
  getCachedOAuthToken,
  setCachedOAuthToken,
  invalidateCachedOAuthToken,
  decodeJWT,
} from './oauth.js';
import {globalAuthMiddlewareRegistry, applyAuthRequestMiddleware, applyAuthResponseMiddleware} from './middleware.js';

/**
 * Configuration for JWT Bearer authentication.
 */
export interface JwtOAuthConfig {
  /** OAuth client ID */
  clientId: string;
  /** Path to JWT certificate file (cert.pem) */
  certPath: string;
  /** Path to JWT private key file (key.pem) */
  keyPath: string;
  /** Optional passphrase for encrypted private key */
  passphrase?: string;
  /** Account Manager hostname */
  accountManagerHost: string;
  /** OAuth scopes to request */
  scopes?: string[];
}

interface JwtPayload {
  iss: string;
  sub: string;
  aud: string;
  exp: number;
}

interface JwtHeader {
  alg: string;
  typ: string;
}

/**
 * OAuth 2.0 JWT Bearer authentication strategy.
 *
 * Implements RFC 7523 (JSON Web Token (JWT) Profile for OAuth 2.0 Client
 * Authentication and Authorization Grants).
 *
 * Key differences from client credentials flow:
 * - Uses public/private key pair instead of client secret
 * - Sends JWT as `client_assertion` in POST body (not Authorization header)
 * - JWT is self-signed and short-lived (60 seconds)
 *
 * @example
 * ```typescript
 * const strategy = new JwtOAuthStrategy({
 *   clientId: 'my-client-id',
 *   certPath: './cert.pem',
 *   keyPath: './key.pem',
 *   accountManagerHost: 'account.demandware.com',
 * });
 *
 * const response = await strategy.fetch('https://api.example.com/data');
 * ```
 */
export class JwtOAuthStrategy implements AuthStrategy {
  private readonly config: JwtOAuthConfig;
  private readonly logger = getLogger();
  private readonly cacheKey: string;
  private _hasHadSuccess = false;
  private readonly privateKey: crypto.KeyObject;

  constructor(config: JwtOAuthConfig) {
    this.validateConfig(config);
    this.config = config;
    this.cacheKey = getOAuthCacheKey(this.config.clientId, 'jwt', this.config.accountManagerHost, this.config.scopes);

    // Cache private key to avoid file I/O on every token request
    const keyContent = fs.readFileSync(config.keyPath, 'utf8');
    this.privateKey = crypto.createPrivateKey({
      key: keyContent,
      passphrase: config.passphrase,
    });
  }

  /**
   * Validates JWT configuration and checks that certificate/key files exist and are readable.
   */
  private validateConfig(config: JwtOAuthConfig): void {
    // Validate required fields
    if (!config.clientId) {
      throw new Error('JWT authentication requires clientId');
    }
    if (!config.certPath) {
      throw new Error('JWT authentication requires certificate path (--jwt-cert)');
    }
    if (!config.keyPath) {
      throw new Error('JWT authentication requires private key path (--jwt-key)');
    }
    if (!config.accountManagerHost) {
      throw new Error('JWT authentication requires accountManagerHost');
    }

    // Validate certificate file exists and has content
    if (!fs.existsSync(config.certPath)) {
      throw new Error(
        `JWT certificate file not found: ${config.certPath}\n` +
          `Generate a certificate pair with: openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365 -nodes`,
      );
    }

    // Check certificate file is readable
    try {
      const certContent = fs.readFileSync(config.certPath, 'utf8');
      if (!certContent.includes('BEGIN CERTIFICATE')) {
        throw new Error(`Invalid certificate format in ${config.certPath}. Expected PEM format (BEGIN CERTIFICATE).`);
      }
    } catch (error) {
      if (error instanceof Error && error.message.includes('Invalid certificate format')) {
        throw error;
      }
      throw new Error(`Failed to read JWT certificate from ${config.certPath}: ${error}`);
    }

    // Validate key file exists
    if (!fs.existsSync(config.keyPath)) {
      throw new Error(
        `JWT private key file not found: ${config.keyPath}\n` +
          `Generate a key pair with: openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365 -nodes`,
      );
    }

    // Validate private key is readable and valid
    try {
      const keyContent = fs.readFileSync(config.keyPath, 'utf8');

      // Check key format
      if (!keyContent.includes('BEGIN') || !keyContent.includes('PRIVATE KEY')) {
        throw new Error(
          `Invalid private key format in ${config.keyPath}. Expected PEM format (BEGIN PRIVATE KEY or BEGIN RSA PRIVATE KEY).`,
        );
      }

      // Validate key can be loaded (will throw if encrypted and passphrase is wrong)
      crypto.createPrivateKey({
        key: keyContent,
        passphrase: config.passphrase,
      });
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);

      // Specific error for wrong passphrase
      if (message.includes('bad decrypt') || message.includes('wrong passphrase') || message.includes('incorrect')) {
        throw new Error(
          `Invalid passphrase for encrypted JWT private key.\n` +
            `Use --jwt-passphrase flag or SFCC_JWT_PASSPHRASE environment variable to provide the passphrase.`,
        );
      }

      // Specific error for encrypted key without passphrase
      if (message.includes('encrypted') || message.includes('passphrase')) {
        throw new Error(
          `JWT private key is encrypted but no passphrase provided.\n` +
            `Use --jwt-passphrase flag or SFCC_JWT_PASSPHRASE environment variable.`,
        );
      }

      // Specific error for invalid format
      if (message.includes('Invalid private key format')) {
        throw error;
      }

      // Generic error
      throw new Error(`Invalid JWT private key at ${config.keyPath}: ${message}`);
    }
  }

  /**
   * Performs a fetch request with JWT Bearer authentication.
   * Automatically injects the Authorization header with a fresh access token.
   * Includes 401 retry logic and x-dw-client-id header.
   */
  async fetch(url: string, init: FetchInit = {}): Promise<Response> {
    const token = await this.getAccessToken();

    const headers = new Headers(init.headers);
    headers.set('Authorization', `Bearer ${token}`);
    headers.set('x-dw-client-id', this.config.clientId);

    // Pass through dispatcher for TLS/mTLS support
    let res = await fetch(url, {...init, headers} as RequestInit);

    if (res.status !== 401) {
      this._hasHadSuccess = true;
    }

    // If we previously had a successful response and now get a 401,
    // the token likely expired. Retry once after invalidating the cached token.
    // Skip retry on initial 401 to avoid retrying with bad credentials.
    if (res.status === 401 && this._hasHadSuccess) {
      this.invalidateToken();
      const newToken = await this.getAccessToken();
      headers.set('Authorization', `Bearer ${newToken}`);
      res = await fetch(url, {...init, headers} as RequestInit);
    }

    return res;
  }

  /**
   * Returns the Authorization header value for legacy clients.
   */
  async getAuthorizationHeader(): Promise<string> {
    const token = await this.getAccessToken();
    return `Bearer ${token}`;
  }

  /**
   * Gets the decoded JWT payload.
   */
  async getJWT(): Promise<ReturnType<typeof decodeJWT>> {
    const token = await this.getAccessToken();
    return decodeJWT(token);
  }

  /**
   * Creates a new JwtOAuthStrategy with additional scopes merged in.
   * Used by clients that have specific scope requirements.
   *
   * @param additionalScopes - Scopes to add to this strategy's existing scopes
   * @returns A new JwtOAuthStrategy instance with merged scopes
   */
  withAdditionalScopes(additionalScopes: string[]): JwtOAuthStrategy {
    const mergedScopes = [...new Set([...(this.config.scopes || []), ...additionalScopes])];
    return new JwtOAuthStrategy({
      ...this.config,
      scopes: mergedScopes,
    });
  }

  /**
   * Gets the full token response including expiration and scopes.
   * Useful for commands that need to display or return token metadata.
   */
  async getTokenResponse(): Promise<AccessTokenResponse> {
    const cached = getCachedOAuthToken(this.cacheKey, this.config.scopes || []);

    if (cached) {
      this.logger.debug('[JwtOAuthStrategy] Reusing cached access token');
      return cached;
    }

    // Get new token (returns full response)
    return this.requestNewToken();
  }

  /**
   * Invalidates the cached access token, forcing re-authentication on next request.
   */
  invalidateToken(): void {
    invalidateCachedOAuthToken(this.cacheKey);
    this.logger.trace('[JwtOAuthStrategy] Token invalidated');
  }

  /**
   * Gets an access token string, using cached token if still valid.
   */
  private async getAccessToken(): Promise<string> {
    // Check global cache first
    const cached = getCachedOAuthToken(this.cacheKey, this.config.scopes || []);
    if (cached) {
      this.logger.trace('[JwtOAuthStrategy] Using cached access token from global cache');
      return cached.accessToken;
    }

    // Request new token and return just the access token string
    const tokenResponse = await this.requestNewToken();
    return tokenResponse.accessToken;
  }

  /**
   * Requests a new access token from Account Manager using JWT Bearer flow.
   * Returns the full token response and caches it.
   */
  private async requestNewToken(): Promise<AccessTokenResponse> {
    this.logger.trace('[JwtOAuthStrategy] Requesting new access token with JWT Bearer flow');

    // Generate signed JWT
    const jwt = await this.createSignedJwt();

    // Request access token using JWT Bearer flow
    const tokenUrl = `https://${this.config.accountManagerHost}/dwsso/oauth2/access_token`;

    // IMPORTANT: JWT credentials go in POST body, NOT Authorization header
    // This is the key difference from client_credentials flow
    const params = new URLSearchParams({
      grant_type: 'client_credentials',
      client_assertion_type: 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer',
      client_assertion: jwt, // ← JWT in body, not header
    });

    if (this.config.scopes && this.config.scopes.length > 0) {
      params.append('scope', this.config.scopes.join(' '));
    }

    this.logger.trace(
      {
        tokenUrl,
        clientId: this.config.clientId,
        scopes: this.config.scopes,
      },
      '[JwtOAuthStrategy] Sending JWT Bearer token request',
    );

    // Build request object for middleware
    let request = new Request(tokenUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: params.toString(),
    });

    // Apply auth middleware (e.g., User-Agent)
    const middleware = globalAuthMiddlewareRegistry.getMiddleware();
    request = await applyAuthRequestMiddleware(request, middleware);

    let response = await fetch(request);

    // Apply auth response middleware
    response = await applyAuthResponseMiddleware(request, response, middleware);

    if (!response.ok) {
      const errorText = await response.text();
      this.logger.error(
        {
          status: response.status,
          statusText: response.statusText,
          error: errorText,
        },
        '[JwtOAuthStrategy] JWT authentication failed',
      );

      // Provide helpful error messages for common issues
      if (response.status === 401) {
        throw new Error(
          `JWT authentication failed (401): Invalid JWT signature or unregistered certificate. ` +
            `Ensure the certificate (${this.config.certPath}) is registered in Account Manager.`,
        );
      }
      if (response.status === 400) {
        throw new Error(`JWT authentication failed (400): ${errorText}`);
      }

      throw new Error(`JWT authentication failed: ${response.status} ${response.statusText}\n${errorText}`);
    }

    const data = (await response.json()) as {access_token?: string; expires_in?: number};
    if (!data.access_token) {
      throw new Error('No access token in response from Account Manager');
    }

    // Calculate token expiry (default 30 minutes if not specified)
    const expiresInSeconds = data.expires_in ?? 1800;
    const expiryDate = new Date(Date.now() + expiresInSeconds * 1000);

    // Decode JWT to extract scopes (scope can be string or array)
    const decoded = decodeJWT(data.access_token);
    const scope = decoded.payload.scope as string | string[] | undefined;
    const scopes = Array.isArray(scope) ? scope : scope?.split(' ') || this.config.scopes || [];

    // Build and cache token response
    const tokenResponse: AccessTokenResponse = {
      accessToken: data.access_token,
      expires: expiryDate,
      scopes,
    };
    setCachedOAuthToken(this.cacheKey, tokenResponse);

    this.logger.trace(
      {
        expiresIn: expiresInSeconds,
        expiresAt: expiryDate.toISOString(),
        scopes,
      },
      '[JwtOAuthStrategy] Access token obtained successfully',
    );

    return tokenResponse;
  }

  /**
   * Creates and signs a JWT token for OAuth authentication.
   * Uses RS256 algorithm and Base64URL encoding per RFC 7519.
   */
  private async createSignedJwt(): Promise<string> {
    const header: JwtHeader = {
      alg: 'RS256',
      typ: 'JWT',
    };
    const encodedHeader = base64UrlEncode(JSON.stringify(header));
    const now = Math.floor(Date.now() / 1000);
    const tokenUrl = `https://${this.config.accountManagerHost}/dwsso/oauth2/access_token`;
    const payload: JwtPayload = {
      iss: this.config.clientId,
      sub: this.config.clientId,
      aud: tokenUrl,
      exp: now + 60,
    };
    const encodedPayload = base64UrlEncode(JSON.stringify(payload));

    const signatureInput = `${encodedHeader}.${encodedPayload}`;
    const signature = crypto.sign('RSA-SHA256', Buffer.from(signatureInput), this.privateKey);
    const encodedSignature = base64UrlEncode(signature);

    const jwt = `${encodedHeader}.${encodedPayload}.${encodedSignature}`;

    this.logger.trace(
      {
        header,
        payload: {
          ...payload,
          exp: new Date(payload.exp * 1000).toISOString(),
        },
      },
      '[JwtOAuthStrategy] Generated JWT token',
    );

    return jwt;
  }
}

/**
 * Encodes data as Base64URL (RFC 4648 Section 5).
 * Replaces +/= with URL-safe characters.
 */
function base64UrlEncode(data: string | Buffer): string {
  const buffer = typeof data === 'string' ? Buffer.from(data) : data;
  const base64 = buffer.toString('base64');
  return base64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=/g, '');
}

File: python/b2c-tooling-sdk/src/b2c_tooling_sdk/auth/oauth_jwt.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""JWT Bearer OAuth authentication strategy (RFC 7523).

Mirrors ``src/auth/oauth-jwt.ts``. Implements the OAuth 2.0 JWT Bearer Token flow
for Account Manager: a self-signed, short-lived (60s) RS256 assertion sent as
``client_assertion`` in the POST **body** (not the Authorization header). Uses a
client certificate/key pair instead of a client secret.
"""

from __future__ import annotations

import base64
import json
import time
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Any
from urllib.parse import urlencode

import httpx
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding
from cryptography.hazmat.primitives.asymmetric.rsa import RSAPrivateKey

from b2c_tooling_sdk.auth.dispatch_fetch import dispatch_fetch
from b2c_tooling_sdk.auth.jwt_utils import decode_jwt
from b2c_tooling_sdk.auth.middleware import (
    apply_auth_request_middleware,
    apply_auth_response_middleware,
    global_auth_middleware_registry,
)
from b2c_tooling_sdk.auth.oauth import (
    find_cached_token_satisfying,
    get_cached_oauth_token,
    get_oauth_cache_key,
    invalidate_cached_tokens_for_identity,
    set_cached_oauth_token,
)
from b2c_tooling_sdk.auth.types import AccessTokenResponse, DecodedJWT
from b2c_tooling_sdk.errors.network_error import wrap_network_error
from b2c_tooling_sdk.logging import get_logger

# Default token lifetime (seconds) when AM omits expires_in.
_DEFAULT_EXPIRES_IN = 1800
# JWT assertion lifetime (seconds).
_ASSERTION_LIFETIME = 60


def _base64url_encode(data: bytes) -> str:
    """Encode bytes as unpadded base64url (RFC 4648 §5)."""
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


class JwtOAuthConfig:
    """Configuration for :class:`JwtOAuthStrategy`."""

    def __init__(
        self,
        client_id: str,
        cert_path: str,
        key_path: str,
        account_manager_host: str,
        passphrase: str | None = None,
        scopes: list[str] | None = None,
    ) -> None:
        self.client_id = client_id
        self.cert_path = cert_path
        self.key_path = key_path
        self.account_manager_host = account_manager_host
        self.passphrase = passphrase
        self.scopes = scopes


class JwtOAuthStrategy:
    """OAuth 2.0 JWT Bearer authentication strategy (RFC 7523).

    Differs from client credentials: uses a public/private key pair instead of a
    secret, sends a self-signed short-lived JWT as ``client_assertion`` in the POST
    body, and shares the module-level token cache under the ``jwt`` method.
    """

    def __init__(self, config: JwtOAuthConfig) -> None:
        self._validate_config(config)
        self._config = config
        self._logger = get_logger("auth.oauth_jwt")
        self._cache_key = get_oauth_cache_key(config.client_id, "jwt", config.account_manager_host, config.scopes)
        self._identity_prefix = f"{config.account_manager_host}:{config.client_id}:jwt:"
        self._has_had_success = False
        # Cache the private key to avoid file I/O on every token request.
        self._private_key = self._load_private_key(config.key_path, config.passphrase)

    @staticmethod
    def _load_private_key(key_path: str, passphrase: str | None) -> Any:
        key_bytes = Path(key_path).read_bytes()
        password = passphrase.encode("utf-8") if passphrase else None
        return serialization.load_pem_private_key(key_bytes, password=password)

    def _validate_config(self, config: JwtOAuthConfig) -> None:
        if not config.client_id:
            raise ValueError("JWT authentication requires clientId")
        if not config.cert_path:
            raise ValueError("JWT authentication requires certificate path (--jwt-cert)")
        if not config.key_path:
            raise ValueError("JWT authentication requires private key path (--jwt-key)")
        if not config.account_manager_host:
            raise ValueError("JWT authentication requires accountManagerHost")

        cert = Path(config.cert_path)
        if not cert.exists():
            raise ValueError(
                f"JWT certificate file not found: {config.cert_path}\n"
                "Generate a certificate pair with: "
                "openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365 -nodes"
            )
        cert_content = cert.read_text(encoding="utf-8", errors="replace")
        if "BEGIN CERTIFICATE" not in cert_content:
            raise ValueError(
                f"Invalid certificate format in {config.cert_path}. Expected PEM format (BEGIN CERTIFICATE)."
            )

        key = Path(config.key_path)
        if not key.exists():
            raise ValueError(
                f"JWT private key file not found: {config.key_path}\n"
                "Generate a key pair with: "
                "openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365 -nodes"
            )
        key_content = key.read_text(encoding="utf-8", errors="replace")
        if "BEGIN" not in key_content or "PRIVATE KEY" not in key_content:
            raise ValueError(
                f"Invalid private key format in {config.key_path}. "
                "Expected PEM format (BEGIN PRIVATE KEY or BEGIN RSA PRIVATE KEY)."
            )

        # Validate the key loads (surfacing encrypted-key / wrong-passphrase issues).
        password = config.passphrase.encode("utf-8") if config.passphrase else None
        try:
            serialization.load_pem_private_key(key.read_bytes(), password=password)
        except TypeError as error:
            # cryptography raises TypeError when a password is needed but absent, or given but not needed.
            message = str(error).lower()
            if "encrypted" in message or "password" in message:
                raise ValueError(
                    "JWT private key is encrypted but no passphrase provided.\n"
                    "Use --jwt-passphrase flag or SFCC_JWT_PASSPHRASE environment variable."
                ) from error
            raise ValueError(f"Invalid JWT private key at {config.key_path}: {error}") from error
        except ValueError as error:
            raise ValueError(
                "Invalid passphrase for encrypted JWT private key.\n"
                "Use --jwt-passphrase flag or SFCC_JWT_PASSPHRASE environment variable to provide the passphrase."
            ) from error

    async def fetch(
        self,
        url: str,
        *,
        method: str = "GET",
        headers: dict[str, str] | None = None,
        content: Any = None,
        dispatcher: httpx.AsyncBaseTransport | None = None,
        **kwargs: Any,
    ) -> httpx.Response:
        """Perform a request with JWT Bearer auth, retrying once on a post-success 401."""
        token = await self._get_access_token()
        request_headers = dict(headers or {})
        request_headers["Authorization"] = f"Bearer {token}"
        request_headers["x-dw-client-id"] = self._config.client_id

        response = await dispatch_fetch(
            url, method=method, headers=request_headers, content=content, dispatcher=dispatcher, **kwargs
        )
        if response.status_code != 401:
            self._has_had_success = True
        if response.status_code == 401 and self._has_had_success:
            self.invalidate_token()
            new_token = await self._get_access_token()
            request_headers["Authorization"] = f"Bearer {new_token}"
            response = await dispatch_fetch(
                url, method=method, headers=request_headers, content=content, dispatcher=dispatcher, **kwargs
            )
        return response

    async def get_authorization_header(self) -> str:
        """Return the ``Authorization`` header value (``Bearer <token>``)."""
        token = await self._get_access_token()
        return f"Bearer {token}"

    async def get_jwt(self) -> DecodedJWT:
        """Return the decoded (unverified) access-token JWT."""
        token = await self._get_access_token()
        return decode_jwt(token)

    async def get_token_response(self) -> AccessTokenResponse:
        """Return the full token response, using the cache when valid."""
        cached = get_cached_oauth_token(self._cache_key, self._config.scopes or [])
        if cached is not None:
            self._logger.debug("[JwtOAuthStrategy] Reusing cached access token")
            return cached
        return await self._request_new_token()

    def invalidate_token(self) -> None:
        """Evict every cached token for this client/AM-host JWT identity."""
        invalidate_cached_tokens_for_identity(self._identity_prefix)

    def with_additional_scopes(self, additional_scopes: list[str]) -> JwtOAuthStrategy:
        """Return a new strategy with ``additional_scopes`` merged into the configured scopes."""
        merged = list(dict.fromkeys([*(self._config.scopes or []), *additional_scopes]))
        return JwtOAuthStrategy(
            JwtOAuthConfig(
                client_id=self._config.client_id,
                cert_path=self._config.cert_path,
                key_path=self._config.key_path,
                account_manager_host=self._config.account_manager_host,
                passphrase=self._config.passphrase,
                scopes=merged,
            )
        )

    async def get_access_token_for_cascade(self, candidates: list[list[str]]) -> str:
        """Resolve a scope cascade for the JWT flow (mirrors :meth:`OAuthStrategy.get_access_token_for_cascade`)."""
        base_scopes = self._config.scopes or []
        for candidate in candidates:
            required = list(dict.fromkeys([*base_scopes, *candidate]))
            cached = find_cached_token_satisfying(self._identity_prefix, required)
            if cached is not None:
                self._logger.debug("[JwtOAuthStrategy] Cache hit for cascade candidate %s", candidate)
                return cached.access_token

        last_error: BaseException | None = None
        for candidate in candidates:
            merged = list(dict.fromkeys([*base_scopes, *candidate]))
            try:
                self._logger.debug("[JwtOAuthStrategy] Cascade trying scopes %s", candidate)
                token_response = await self._request_new_token_for_scopes(merged)
                return token_response.access_token
            except Exception as error:  # noqa: BLE001 - invalid_scope expected; others rethrow
                if "invalid_scope" in str(error):
                    self._logger.debug(
                        "[JwtOAuthStrategy] Candidate %s rejected (invalid_scope), trying next", candidate
                    )
                    last_error = error
                    continue
                raise
        if last_error is not None:
            raise last_error
        raise RuntimeError("All scope cascade candidates failed")

    async def _get_access_token(self) -> str:
        cached = get_cached_oauth_token(self._cache_key, self._config.scopes or [])
        if cached is not None:
            return cached.access_token
        token_response = await self._request_new_token()
        return token_response.access_token

    async def _request_new_token(self) -> AccessTokenResponse:
        return await self._request_new_token_for_scopes(self._config.scopes)

    async def _request_new_token_for_scopes(self, scopes: list[str] | None) -> AccessTokenResponse:
        assertion = self._create_signed_jwt()
        token_url = f"https://{self._config.account_manager_host}/dwsso/oauth2/access_token"

        # JWT credentials go in the POST body, NOT the Authorization header.
        params: dict[str, str] = {
            "grant_type": "client_credentials",
            "client_assertion_type": "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
            "client_assertion": assertion,
        }
        if scopes:
            params["scope"] = " ".join(scopes)
        body = urlencode(params)

        headers = {"Content-Type": "application/x-www-form-urlencoded"}
        middleware = global_auth_middleware_registry.get_middleware()

        async with httpx.AsyncClient() as client:
            request = client.build_request("POST", token_url, content=body, headers=headers)
            request = await apply_auth_request_middleware(request, middleware)
            try:
                response = await client.send(request)
            except Exception as err:  # noqa: BLE001 - wrapped only if network-level
                host = httpx.URL(token_url).host
                raise wrap_network_error(err, operation="OAuth JWT token request", host=host) from err
            response = await apply_auth_response_middleware(request, response, middleware)

            if not response.is_success:
                error_text = response.text
                self._logger.error("[JwtOAuthStrategy] JWT authentication failed: %s", response.status_code)
                if response.status_code == 401:
                    raise RuntimeError(
                        f"JWT authentication failed (401): Invalid JWT signature or unregistered certificate. "
                        f"Ensure the certificate ({self._config.cert_path}) is registered in Account Manager."
                    )
                if response.status_code == 400:
                    raise RuntimeError(f"JWT authentication failed (400): {error_text}")
                raise RuntimeError(
                    f"JWT authentication failed: {response.status_code} {response.reason_phrase}\n{error_text}"
                )
            data = response.json()

        access_token = data.get("access_token")
        if not access_token:
            raise RuntimeError("No access token in response from Account Manager")

        expires_in = data.get("expires_in") or _DEFAULT_EXPIRES_IN
        expiry = datetime.now(tz=timezone.utc) + timedelta(seconds=expires_in)

        decoded = decode_jwt(access_token)
        scope = decoded.payload.get("scope")
        if isinstance(scope, list):
            token_scopes = [str(s) for s in scope]
        elif isinstance(scope, str):
            token_scopes = scope.split(" ")
        else:
            token_scopes = scopes or []

        token_response = AccessTokenResponse(access_token=access_token, expires=expiry, scopes=token_scopes)
        cache_key = get_oauth_cache_key(self._config.client_id, "jwt", self._config.account_manager_host, scopes)
        set_cached_oauth_token(cache_key, token_response)
        return token_response

    def _create_signed_jwt(self) -> str:
        header = {"alg": "RS256", "typ": "JWT"}
        encoded_header = _base64url_encode(json.dumps(header).encode("utf-8"))
        now = int(time.time())
        token_url = f"https://{self._config.account_manager_host}/dwsso/oauth2/access_token"
        payload = {
            "iss": self._config.client_id,
            "sub": self._config.client_id,
            "aud": token_url,
            "exp": now + _ASSERTION_LIFETIME,
        }
        encoded_payload = _base64url_encode(json.dumps(payload).encode("utf-8"))
        signature_input = f"{encoded_header}.{encoded_payload}".encode("ascii")
        if not isinstance(self._private_key, RSAPrivateKey):
            raise ValueError("JWT Bearer flow requires an RSA private key")
        signature = self._private_key.sign(signature_input, padding.PKCS1v15(), hashes.SHA256())
        encoded_signature = _base64url_encode(signature)
        return f"{encoded_header}.{encoded_payload}.{encoded_signature}"


__all__ = ["JwtOAuthConfig", "JwtOAuthStrategy"]
